src/PostSharp.Engineering.BuildTools/Dependencies/FetchDependencyCommand.cs

```diff
@@ -77,44 +77,140 @@ protected override bool ExecuteCore( BuildContext context, FetchDependenciesComm
-            if ( !ResolveTransitiveDependencies( context, teamcity, dependencies, versionsOverrideFile ) )
-            {
-                return false;
-            }
+                if ( !DownloadArtifacts( context, teamcity, iterationDependencies ) )
+                {
+                    return false;
+                }
 
-            // Download transitive dependencies.
-            if ( !ResolveBuildNumbersFromBranches( context, teamcity, dependencies, false ) )
-            {
-                return false;
+                // Find implicit transitive dependencies.
+                if ( !TryGetTransitiveDependencies( context, iterationDependencies, versionsOverrideFile, out var allDependencies ) )
+                {
+                    return false;
+                }
+
+                // Resolve transitive dependencies from artefacts that have been downloaded.
+                // (We currently only support a single level of dependencies. To support several levels, this process should be iterative). 
+                if ( !ResolveTransitiveDependencies( context, allDependencies, versionsOverrideFile, out var resolvedDependencies ) )
+                {
+                    return false;
+                }
+
+                iterationDependencies = resolvedDependencies;
             }
 
-            if ( !DownloadArtifacts( context, teamcity, dependencies ) )
+            context.Console.WriteSuccess( "Fetching build artifacts was successful" );
+
+            return true;
+        }
+
+        private static bool TryGetTransitiveDependencies(
+            BuildContext context,
+            List<Dependency> directDependencies,
+            VersionsOverrideFile versionsOverrideFile,
+            [NotNullWhen( true )] out List<Dependency>? allDependencies )
+        {
+            var dependencies = directDependencies.ToDictionary( d => d.Definition.Name, d => d );
+
+            foreach ( var directDependency in directDependencies )
             {
-                return false;
+                if ( directDependency.Source.SourceKind == DependencySourceKind.Transitive )
+                {
+                    // Defining transitive dependencies explicitly is now obsolete, but we can still get here 
+                    // because of a previous iteration.
+                    continue;
+                }
+
+                if ( directDependency.Source.SourceKind == DependencySourceKind.Default )
+                {
+                    // There is no need to resolve dependencies deployed to a public source because nuget does it.
+                    continue;
+                }
+
+                var versionFile = Project.FromFile( directDependency.Source.VersionFile!, new ProjectOptions() );
+
+                var transitiveDependencies = versionFile.Items.Where( i => i.ItemType == "MetalamaDependencies" );
```

Comment:
The item type name depends on the product name. (In Metalama.version.props, it is "MetalamaDependencies", in Metalama.Compiler.version.props, it is MetalamaCompilerDependencies, etc.) It is so because all the props files can be imported into one MSBuild project, which would mix all the items from all reports. (However, it is not important here, because we load each .versions.props file separately.)